File: triton-bot/src/main/python/module/module.py
import pickle
from threading import Thread

from google.protobuf.message import Message
from constant.runtime_constants import RuntimeConstants
import pika

class Module(Thread):
    CONNECTION_FACTORY_HOST = 'localhost'
    FANOUT = 'fanout'

    def __init__(self):
        super().__init__()
        self.setup_channels()
        self.load_config()
        self.prepare()
        self.declare_publishes()
        self.declare_consumes()

    def setup_channels(self):
        self.consume_connection = pika.BlockingConnection(
            pika.ConnectionParameters(Module.CONNECTION_FACTORY_HOST, heartbeat=10))
        self.consume_channel = self.consume_connection.channel()

        self.publish_connection = pika.BlockingConnection(
            pika.ConnectionParameters(Module.CONNECTION_FACTORY_HOST, heartbeat=10))
        self.publish_channel = self.publish_connection.channel()

    def load_config(self):
        pass

    def prepare(self):
        pass

    def declare_publishes(self):
        pass

    def declare_consumes(self):
        pass

    def declare_publish(self, exchange):
        self.publish_channel.exchange_declare(
            exchange=exchange.name + str(RuntimeConstants.team) + str(RuntimeConstants.id), exchange_type=Module.FANOUT)

    def declare_consume(self, exchange, callback):
        self.consume_channel.exchange_declare(
            exchange=exchange.name + str(RuntimeConstants.team) + str(RuntimeConstants.id), exchange_type=Module.FANOUT)

        queue_name = self.consume_channel.queue_declare(
            queue='', exclusive=True).method.queue
        self.consume_channel.queue_bind(exchange=exchange.name + str(RuntimeConstants.team) + str(RuntimeConstants.id), queue=queue_name)

        self.consume_channel.basic_consume(
            queue=queue_name, on_message_callback=callback, auto_ack=True)

    def publish(self, exchange, object):
        if (isinstance(object, Message)):
            body = object.SerializeToString()
        else:
            body = pickle.dumps(object)

        self.publish_channel.basic_publish(
            exchange=exchange.name + str(RuntimeConstants.team) + str(RuntimeConstants.id), routing_key='', body=body)
    
    def consume(self):
        self.consume_channel.start_consuming()

File: triton-soccer-ai/src/main/java/com/triton/module/Module.java
package com.triton.module;

import com.rabbitmq.client.*;
import com.triton.constant.ProgramConstants;
import com.triton.messaging.Exchange;

import java.io.IOException;
import java.util.Date;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.ScheduledThreadPoolExecutor;
import java.util.concurrent.TimeoutException;

import static com.triton.messaging.SimpleSerialize.simpleSerialize;

public abstract class Module extends Thread {
    private static final String CONNECTION_FACTORY_HOST = "localhost";
    private static final String FANOUT = "fanout";

    public ScheduledThreadPoolExecutor executor;

    private ConnectionFactory factory;
    private Channel publish_channel;
    private Channel consume_channel;

    public Module(ScheduledThreadPoolExecutor executor) {
        this.executor = executor;
        try {
            setupChannel();
            prepare();
            declareConsumes();
        } catch (IOException | TimeoutException e) {
            e.printStackTrace();
        }
    }

    private void setupChannel() throws IOException, TimeoutException {
        factory = new ConnectionFactory();
        factory.setHost(CONNECTION_FACTORY_HOST);

        Connection publish_connection = factory.newConnection();
        publish_channel = publish_connection.createChannel();

        Connection consume_connection = factory.newConnection();
        consume_channel = consume_connection.createChannel();
    }

    protected abstract void prepare();

    protected abstract void declareConsumes() throws IOException, TimeoutException;

    /**
     * Declares an exchange to consume from. The messageConsumer function will be called when an message is received
     * from the exchange.
     *
     * @param exchange the exchange to consume from
     * @param callback the function to call once a message is received
     * @throws IOException
     */
    public void declareConsume(Exchange exchange, DeliverCallback callback) throws IOException, TimeoutException {
        consume_channel.exchangeDeclare(exchange.name() + ProgramConstants.team.name(), FANOUT);

        Map<String, Object> args = new HashMap<>();
        args.put("x-message-ttl", 1000);
        args.put("x-expires", 1000);
        String queueName = consume_channel.queueDeclare("",
                false,
                false,
                false,
                args).getQueue();
        consume_channel.queueBind(queueName, exchange.name() + ProgramConstants.team.name(), "");
        consume_channel.queuePurge(queueName);

        DeliverCallback wrappedCallback = (s, delivery) -> {
            try {
//                long timeDiff = new Date().getTime() - delivery.getProperties().getTimestamp().getTime();
//                if (timeDiff > 10000) {
//                    System.out.println("RabbitMQ Bottleneck Warning: " + timeDiff + " ms");
//                    System.out.println("Class: " + this.getClass());
//                    System.out.println("Exchange: " + exchange.name());
//                } else {
                callback.handle(s, delivery);
//                }
            } catch (Exception e) {
                e.printStackTrace();
            }
        };
        consume_channel.basicConsume(queueName, true, wrappedCallback, consumerTag -> {
        });
    }

    /**
     * Publishes to an exchange
     *
     * @param exchange the exchange to publish to
     * @param object   the object to send
     * @throws IOException
     */
    public void publish(Exchange exchange, Object object) {
        Date date = new Date();
        AMQP.BasicProperties properties = new AMQP.BasicProperties.Builder().timestamp(date).build();
        if (publish_channel.isOpen()) {
            try {
                declarePublish(exchange);
                publish_channel.basicPublish(exchange.name() + ProgramConstants.team.name(), "", properties, simpleSerialize(object));
            } catch (IOException | AlreadyClosedException ignored) {
            }
        }
    }

    /**
     * Declares an exchange to publish to
     *
     * @param exchange the exchange
     * @throws IOException
     */
    public void declarePublish(Exchange exchange) throws IOException {
        publish_channel.exchangeDeclare(exchange.name() + ProgramConstants.team.name(), FANOUT);
    }

    @Override
    public void interrupt() {
        super.interrupt();
        try {
            consume_channel.close();
            publish_channel.close();
        } catch (IOException | TimeoutException e) {
            e.printStackTrace();
        }
    }
}
